### PR TITLE
[Feature] Quasar prefill SDPA: functional on craq-sim + emulator

### LINKED ISSUE #56689: [Quasar] Bring up experimental sdpa (prefill) on Quasar (functional on craq-sim + emulator + WH)
Tracking the bring-up of the prefill `ttnn.experimental.quasar.transformer.scaled_dot_product_attention` fork (`ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/`) from compile-only to numerically-correct on Quasar. **Scope: prefill `sdpa` only** — the `sdpa_decode` fork is tracked separately in #56688.

Follows the Metal 2.0 port (#55218, closed) and the unpack_modes variant fix (#55733, closed). Tests: `models/experimental/llama32_1b_quasar/tests/ops/test_scaled_dot_product_attention.py`. Fixes are referenced by commit SUBJECT — the branch is mid-rebase and SHAs are volatile.

## Status (all green)
| Platform | Coverage | Result |
|---|---|---|
| **Wormhole n150** (baseline) | 13/13: ops 3 (seq128 / seq512=4-chunk / seq1024=8-chunk) + ops-chunked 3 + prototype_ops 3 + prototype_ops-chunked 3 + graph 1 | ✅ 13/13 |
| **Emulator** (real Quasar) | seq128 + seq512 + seq1024 | ✅ pass |
| **craq-sim** (Quasar) | seq128 + seq512 + seq1024 (seq1024 ~116s) | ✅ pass |

Branch: `gchoudhary/quasar/get-sdpa-sdpa_decode-functional-on-sim-or-emu` (shared with the decode fork). craq-sim runs on `nkapre/q389-pacr-stride` (libttsim). Note: `test_attention_1d` module test is 9/10 — the one failure is an HF gated-repo auth (OSError), not the op.

**Multi-chunk is directly validated on WH silicon** (seq512 = 4 K-chunks, seq1024 = 8) — prefill has no tree reduction, so unlike the decode fork there is no "multi-chunk only on sim" caveat here.

## Fixes landed (by category, commit subjects)
**Compile / LLK-compat on Quasar**
- Compile + run on craq-sim: swap fused experimental SFPU softmax primitives for generic Quasar SFPU tile ops, streaming-branch guard, fp32-dest gate
- (Prereq: unpack_modes std::get / Gen1-config `bad_variant_access` fix, #55733)

**Intra-Tensix DFB tile-counter budget (Quasar cap = 8) — final state = 7 DFBs, native transpose, SUM merged, MAX split**
- Transpose-free QK^T via a dedicated `kt` DFB (→ 9, over cap) → merge the **SUM** ping-pong into one depth-2 DFB (→ 8); documented the shared merged-sum fma decision rule
- Then **drop the `kt` staging DFB** and use the matmul's native SrcA transpose for QK^T (→ **7**), now that the srca-transpose LLK provides it
- MAX stays **split** (max_A/max_B) — a MAX-merge was tried and **reverted** (multi-chunk-broken)

**Quasar packer operand retargeting** (WH re-targets by named buffer; Quasar needs explicit `pack_init` at each output-operand switch)
- `pack_init` after each output-DFB switch (all-zeros fix), then gated to fire only on a real pack-operand CHANGE (a blanket `pack_init` resets the L1-accumulate section at same-operand reconfigs)

**Multi-chunk online-softmax correctness**
- **The load-bearing multi-chunk fix:** add the missing `pack_reconfig_out(exp_max_diff)` before the correction `sub_exp_block` (a bare `pack_tile` left the packer stuck on the prior QK@V ring → garbage EMD compounding per chunk; single-chunk seq128 passed because it has no correction stage)
- Also present but **not** the fix: a two-op MM2 output rescale (subject "two-op MM2 rescale, not fused L1-accum") — proven bit-identical/inert vs the fused form (EMD was garbage upstream); kept only as decode-form alignment, not load-bearing

## Dependencies
- craq-sim dense `PACR_STRIDE` modeling + the unguarded pack-store OOB fix (tenstorrent/craq-sim#389, #390) — the multi-chunk `cur_sum` reduce SIGBUS'd on readback before it (out-of-bounds pack store), then landed correct bytes after.
- The srca-transpose-matmul LLK — enables the `kt` DFB removal.

## Open (non-blocking)
1. Secondary `attention_sink` instance of the same bare-`pack_tile` class (~`compute_common.hpp:2310`, a `pack_reconfig_out(exp_max_diff)`-style site) — still needs the same treatment; untested on Quasar.
2. JointSDPA untested on Quasar — no joint unit test exists; validated only by host-compile + mirror-of-`sdpa` construction (kt was also removed from the joint factory).
3. Optional cleanup: the SUM-merge is no longer needed for the cap now that `kt` is gone (7/8) — could un-merge to plain `sum_A`/`sum_B` for simplicity. Not a bug.

## Related
- #55218 — Metal 2.0 sdpa port (precursor, closed)
- #55733 — unpack_modes Gen1-config `bad_variant_access` on Quasar (closed)
- #56688 — the sibling `sdpa_decode` Quasar bring-up (separate)
- tenstorrent/craq-sim#389, #390 — the companion simulator fixes

### PR DESCRIPTION
**Pull request by gajanan-choudhary**
### PR Category
Feature

### Summary
Closes #56689.

Brings up the prefill `scaled_dot_product_attention` (experimental Quasar fork) so it runs correctly on Quasar — craq-sim and the ZeBu emulator — while staying green on Wormhole.

Base is the `sdpa_decode` Quasar branch (stacked), not `main`, so this PR is only the prefill-specific delta on top of decode.

Changes (all under `experimental/quasar/transformer/sdpa/`):
- Quasar compile path: generic SFPU (swap the fused softmax primitives), streaming-path compile guard, fp32-dest-acc off.
- QK^T: native matmul SrcA transpose. The `kt`-DFB staging workaround was added while the Quasar matmul unpacker lacked SrcA transpose, then removed once the LLK support landed.
- DFB budget: merge the running-sum ping-pong into one DFB to stay within Quasar's 8 compute self-loop DFBs.
- Packer (Quasar sticky-dest): `pack_init` only on a real output-operand change, and `pack_reconfig_out(exp_max_diff)` before the correction `sub_exp_block` — the load-bearing multi-chunk fix.

Validation: WH 13/13 (ops/chunked/prototype/graph, seq128/512/1024), Quasar emulator seq128/512/1024, craq-sim seq128/512/1024.

### Notes for reviewers
- The one correctness-critical fix is `pack_reconfig_out(exp_max_diff)` before the correction `sub_exp_block`: that helper packs via a bare `pack_tile`, so on Quasar's sticky-dest packer the caller must re-point the packer first (WH re-derives the dest, so WH never surfaced the bug — it only showed up multi-chunk on the emulator).
- A couple of commits are superseded/inert and could be squashed: the kt add→remove pair cancels out; the two-op MM2 rescale is bit-identical to the fused form (kept only to mirror decode). Happy to squash if preferred.
- Depends on the shared LLK SrcA-transpose + DFB infra (authored by others; those commits were dropped from this branch and will be merged independently) — this PR builds on top of that base.
- Not in this PR (still open on Quasar): JointSDPA factory gates, attention-sink path (blocked earlier by a TRISC code-size overflow at -O3), chunked on craq-sim, streaming/ring ports.

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/device/kernels/compute/compute_common.hpp [hunk 1]
```diff
@@ -30,23 +30,78 @@
...
         transpose, true /*transpose within 16x16 face*/, dfbid)));
 
+#ifdef ARCH_QUASAR
+    // Quasar's llk_math_eltwise_unary_datacopy_init takes no PackMode param; its trailing template
+    // args are bools (unpack_to_dest, is_int_fpu_en, tilize).
+    MATH((llk_math_eltwise_unary_datacopy_init<
+          DataCopyType::A2D,
+          DST_ACCUM_MODE,
+          BroadcastType::NONE,
+          false /*unpack_to_dest*/,
```
[the remark this answers is not in the document]
Comment (gajanan-choudhary) [reply ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/device/kernels/compute/compute_common.hpp:95]: The direct LLK call is intentional and documented (comment just above): sdpa_reduce_copy_tile_to_dst_init_short needs a transpose-within-16x16-face datacopy init, and Quasar's llk_math_eltwise_unary_datacopy_init has a different signature (no PackMode param) than WH/BH — hence the arch-gated branches. The public copy_init in tile_move_copy.h doesn't expose the transpose-within-face option this helper requires, so using it would mean adding a new transpose-capable variant to the shared compute API — an infra change to tile_move_copy.h, out of scope for this experimental-op PR. Leaving as-is.

File: ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/device/kernels/compute/compute_common.hpp [hunk 2]
```diff
@@ -833,6 +963,89 @@ void move_block(uint32_t in_dfb, uint32_t out_dfb, uint32_t num_tiles) {
...
+    d_emd.wait_front(num_tiles);
+
+    // Reserve the running-sum region behind cur (buffer depth is 3*num_tiles).
+    d_sum.reserve_back(num_tiles);
+
+    // Pass 1: reserved_running[i] = cur[i].
+    reconfig_data_format_srca(dfb_sum);
+    copy_init(dfb_sum);
+    pack_reconfig_out(dfb_sum);
+    PACK((llk_pack_reconfig_l1_acc(0)));
```
[the remark this answers is not in the document]
Comment (gajanan-choudhary) [reply ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/device/kernels/compute/compute_common.hpp:1014]: pack_reconfig_l1_acc (pack.h:213) is a 1:1 alias of PACK((llk_pack_reconfig_l1_acc(...))), so this is behavior-neutral. This file is a vendored copy of the main-tree compute_common.hpp and uses the direct PACK((llk_pack_reconfig_l1_acc(...))) pattern in 15+ places (matching the main-tree source); the three flagged sites are a subset. Switching only those is inconsistent, and switching all would diverge the vendored fork from main for zero functional change. Deferring as a style nit — better as a consistent sweep in the main-tree header if wanted.

File: ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/device/kernels/compute/compute_common.hpp [hunk 3]
```diff
@@ -2001,6 +2275,18 @@ void sdpa_inner_loop(
...
+         *
+         * NOTE (Quasar bring-up): the attention-sink path below is NOT brought up on Quasar and is
+         * currently unverified there. Llama (the model driving this fork's bring-up) has no attention
+         * sinks, so nothing exercises this path today; it is a GPT-OSS-class feature. Two known gaps
+         * remain, to be resolved when a sink-using model is ported to this op:
+         *   1. Code size: with USE_ATTENTION_SINK=1 the compute kernel overflows the Quasar TRISC
+         *      instruction-memory region at -O3 (~32KB kernel vs ~24KB limit), so it does not even
+         *      load. Needs an Os/code-size fix for the sink path before it can run on craq-sim.
+         *   2. Packer retargeting: the bare-pack helpers in this block (e.g. the sub_exp_block below)
+         *      lack a preceding pack_reconfig_out to point the Quasar packer at their output DFB, the
```
[the remark this answers is not in the document]
Comment (gajanan-choudhary) [reply ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/device/kernels/compute/compute_common.hpp:2285]: Declining to add this guard in this PR — deferring it to the attention-sink Quasar bring-up (a tracked follow-up). Reasoning:

- The general principle (an argument that can't work should fail at validation, not device load) is fair, but the premise here is narrower than stated. On Quasar, attention_sink currently fails LOUDLY at program build, not silently: the sink compute kernel overflows the TRISC instruction-memory region (trisc*.elf segment overflows region limit) -> TT_THROW 'Failed to generate binaries for sdpa' -> RuntimeError. The 'silently writes correction data to the wrong DFB' path (the missing packer retarget) is not reachable today because the kernel cannot fit/load in the first place. So there is no silent-wrong-output risk in the current state; the suggested guard would only convert one fail-fast (build throw) into another (validate throw).

- attention_sink is validated and works on WH/BH, so any guard would have to be Quasar-only, not a blanket reject. This is an experimental Quasar-fork op (ttnn.experimental.quasar.transformer.*) mid-bring-up: several paths (streaming, ring, chunked, sink) are intentionally not-yet-enabled on Quasar and none are individually host-gated. Singling out attention_sink here would be inconsistent.

- The correct place for a sink-on-Quasar validation decision is the sink bring-up itself, which must first resolve the TRISC code-size overflow and the sink-path packer retarget. A reject-guard added here would just be deleted there. Tracking it with the rest of the Quasar sink work rather than pre-emptively rejecting in this standard-path PR.

File: ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/device/kernels/compute/compute_common.hpp [hunk 4]
```diff
@@ -1957,31 +2208,54 @@ void sdpa_inner_loop(
                  * dfb_exp_max_diff = torch.exp((dfb_prev_max - dfb_cur_max) * scale)
                  * Scale is fused into exp again since max is the max of unscaled scores.
                  */
+                // Re-point the packer to the EMD output BEFORE sub_exp_block, which packs via a bare
+                // pack_tile (no internal pack_reconfig_out). On Quasar the packer is sticky-dest and is
+                // still latched to alias_mm2_cur_out from the QK@V matmul above (line 2211 + the matmul
+                // that inherits it); without this the EMD is written to the OUT_IM ring instead of
+                // dfb_exp_max_diff, so every downstream correction consumer (fma sum merge, MM2 rescale)
+                // reads garbage → multi-chunk PCC ~0.03 while seq128 (no correction) passes. Line 2230
+                // above only reconfigures the UNPACKers. Mirrors sdpa_decode's sub_exp_block preamble.
```
[the remark this answers is not in the document]
Comment (gajanan-choudhary) [reply ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/device/kernels/compute/compute_common.hpp]: Fixed in ea678d13ed6 — trimmed to the invariant (bare pack_tile helper -> the Quasar caller must select the pack operand); dropped the hardcoded line numbers and bring-up history.

File: ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/device/kernels/compute/sdpa.cpp [hunk 5]
```diff
@@ -90,7 +96,15 @@ void kernel_main() {
     constexpr auto dfb_max_A = dfb::max_A;
     constexpr auto dfb_max_B = dfb::max_B;
     constexpr auto dfb_sum_A = dfb::sum_A;
+#if defined(USE_STREAMING_COMPUTE) || defined(USE_ATTENTION_SINK)
+    // Streaming and the attention-sink path keep two separate sum DFBs (prev/cur ping-pong halves).
     constexpr auto dfb_sum_B = dfb::sum_B;
+#else
+    // The non-streaming STANDARD path (no sink) merges the running-sum ping-pong into a single 2-deep
```
[the remark this answers is not in the document]
Comment (gajanan-choudhary) [reply ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/device/kernels/compute/sdpa.cpp]: Fixed in ea678d13ed6 — comment now reads 3-deep, matching the SUM_A = 3*statistics_tiles allocation.

File: ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/device/sdpa_program_factory.cpp [hunk 6]
```diff
@@ -687,6 +695,27 @@ ttnn::device_operation::ProgramArtifacts SDPAOperation::SDPAProgramFactory::crea
         DataflowBufferSpec{
             .unique_id = OUT, .entry_size = out_tile_size, .num_entries = out0_t, .data_format_metadata = out_df},
     };
+    // The non-streaming STANDARD path (no attention sink) merges the running-sum ping-pong (SUM_A/
+    // SUM_B) into a single 2-deep DFB bound to SUM_A (depth 2*statistics_tiles): the kernel keeps prev
```
[the remark this answers is not in the document]
Comment (gajanan-choudhary) [reply ttnn/cpp/ttnn/operations/experimental/quasar/transformer/sdpa/device/sdpa_program_factory.cpp]: Fixed in ea678d13ed6 — comment now reads 3-deep / 3*statistics_tiles, matching the allocation.

**Review by bbradelTT — APPROVED** [review]
